04/app.ts

```diff
@@ -0,0 +1,66 @@
+abstract class PluginNew {
+    private static activePlugins: PluginNew[] = [];
+
+    constructor() {
+        this.install();
+    }
+
+    install(): void {
+        PluginNew.activePlugins.push(this);
+    }
+
+    uninstall(): void {
+        PluginNew.activePlugins = PluginNew.activePlugins.filter(plugin => plugin !== this);
+    }
+
+    static showActivePlugins(): void {
+        PluginNew.activePlugins.forEach((plugin) => {
+            if (isLogger(plugin)) { 
+                plugin.print();
+            }
+        });
+    }
+}
+
+interface Logger extends PluginNew {
+    add(message: string): void;
+    print(): void;
+}
+
+function isLogger(plugin: PluginNew): plugin is Logger {
+    return "print" in plugin;
+}
```

Comment:
Pytanie czy potrzebujemy to sprawdzać. Jeśli klasy implementują interfejs Logger to musza posiadać metody `add()` i `print()`. TS nas poinformuje jak tak nie będzie - dlatego nie musimy tego dodatkowo sprawdzać jeśli ten interfejs dodamy do klasy abstrakcyjnej.